File: examples/sidecar-js/decision-analytics/test/anomaly-detector.test.ts
import { describe, expect, it } from 'vitest';

import { AnomalyDetector } from '../src/anomaly-detector.js';

describe('AnomalyDetector', () => {
  it('does not flag normal latency', () => {
    const detector = new AnomalyDetector(20, 3);
    const latencies = Array.from({ length: 10 }, () => 100);
    const result = detector.detect('shipping-rates', latencies, 110);

    expect(result.isAnomaly).toBe(false);
    expect(result.spikeFactor).toBe(1.1);
    expect(result.decisionRef).toBe('shipping-rates');
  });

  it('flags a tenfold spike with correct spike factor', () => {
    const detector = new AnomalyDetector(20, 3);
    const latencies = Array.from({ length: 10 }, () => 100);
    const result = detector.detect('shipping-rates', latencies, 1000);

    expect(result.isAnomaly).toBe(true);
    expect(result.spikeFactor).toBe(10);
    expect(result.rollingAverage).toBe(100);
    expect(result.currentLatency).toBe(1000);
  });

  it('does not flag when history is empty', () => {
    const detector = new AnomalyDetector();
    const result = detector.detect('shipping-rates', [], 5000);

    expect(result.isAnomaly).toBe(false);
    expect(result.spikeFactor).toBe(0);
    expect(result.rollingAverage).toBe(0);
  });

  it('flags at the spike threshold boundary', () => {
    const detector = new AnomalyDetector(20, 3);
    const latencies = [100, 100, 100];
    const result = detector.detect('shipping-rates', latencies, 300);

    expect(result.isAnomaly).toBe(true);
    expect(result.spikeFactor).toBe(3);
  });

  it('does not flag just below the spike threshold', () => {
    const detector = new AnomalyDetector(20, 3);
    const latencies = [100, 100, 100];
    const result = detector.detect('shipping-rates', latencies, 299);

    expect(result.isAnomaly).toBe(false);
    expect(result.spikeFactor).toBe(2.99);
  });
});

File: examples/plugins/business_rules/decision_analytics/test/anomaly_detector_test.exs
defmodule Examples.BusinessRules.DecisionAnalytics.AnomalyDetectorTest do
  use ExUnit.Case, async: true

  alias Examples.BusinessRules.DecisionAnalytics.AnomalyDetector

  test "does not flag normal latency" do
    detector = AnomalyDetector.new(window_size: 20, spike_threshold: 3)
    latencies = List.duplicate(100, 10)
    result = AnomalyDetector.detect(detector, "shipping-rates", latencies, 110)

    refute result.is_anomaly
    assert result.spike_factor == 1.1
    assert result.decision_ref == "shipping-rates"
  end

  test "flags a tenfold spike with correct spike factor" do
    detector = AnomalyDetector.new(window_size: 20, spike_threshold: 3)
    latencies = List.duplicate(100, 10)
    result = AnomalyDetector.detect(detector, "shipping-rates", latencies, 1000)

    assert result.is_anomaly
    assert result.spike_factor == 10.0
    assert result.rolling_average == 100
    assert result.current_latency == 1000
  end

  test "does not flag when history is empty" do
    detector = AnomalyDetector.new()
    result = AnomalyDetector.detect(detector, "shipping-rates", [], 5000)

    refute result.is_anomaly
    assert result.spike_factor == 0.0
    assert result.rolling_average == 0
  end

  test "flags at the spike threshold boundary" do
    detector = AnomalyDetector.new(window_size: 20, spike_threshold: 3)
    result = AnomalyDetector.detect(detector, "shipping-rates", [100, 100, 100], 300)

    assert result.is_anomaly
    assert result.spike_factor == 3.0
  end

  test "does not flag just below the spike threshold" do
    detector = AnomalyDetector.new(window_size: 20, spike_threshold: 3)
    result = AnomalyDetector.detect(detector, "shipping-rates", [100, 100, 100], 299)

    refute result.is_anomaly
    assert result.spike_factor == 2.99
  end
end
